lib/orchestrator.js

```diff
@@ -0,0 +1,354 @@
+ * @returns {Boolean} Whether or not all jobs have finished or one has failed
+ */
+Orchestrator.hasFinished = function (state) {
+  return state.get('status') !== 'started';
+};
+
+/**
+ * Iteratively calls a function over a value until a condition is reached.
+ * This function produces a list of promised values, and a handler function is run on each of these values.
+ *
+ * @param {T => List.<Promise.<T>>} iteration - A function to repeatedly generate a new list of T from a previous T until a condition is reached
+ * @param {T => Boolean} condition - A function that returns true when the iteration must stop for a given T
+ * @param {T => ()} handler - A function to handle each promise produced by `iteration` as soon as they are fulfilled
+ * @param {T} t - A starting value
+ * @returns {Promise.<T>} The result of the last call to `iteration` before it stops
+ */
+Orchestrator.iterate = function (iteration, condition, handler, t) {
+  if (condition(t)) return Promise.resolve(t);
+  else return iteration(t).reduce(function (sequence, promisedNewT) {
+    return sequence.then(function () {
+      return promisedNewT.then(function (newT) {
+        handler(newT);
+        return newT;
+      });
+    });
+  }, Promise.resolve()).then(function (finalT) {
+    return Orchestrator.iterate(iteration, condition, handler, finalT);
+  });
+};
+
+/**
+ * @param {Promise.<Map.<string, *>>} step
+ * @returns {RequestState => List.<Promise.<RequestState>>}
+ */
+Orchestrator.prototype.runStep = function (step) {
+  return function (state) {
+    return List.of(
+      Promise.resolve(state.setJobStatus(step.get('name'), 'pending')),
+      step.get('promise').then(function (obj) {
+        state = state.setJobStatus(step.get('name'), obj.get('status'));
+
+        if (obj.get('status') === 'failure') {
+          state = state.set('status', 'failure');
+        }
+
+        if (obj.has('errors')) {
+          state = state.setJobErrors(step.get('name'), obj.get('errors'));
+        }
+        if (obj.has('history')) state = state.addToHistory(obj.get('history'));
+        if (obj.has('metadata')) {
+          state = state.set('metadata', obj.get('metadata'));
+        }
+
+        return state;
+      })
+      .catch(function (err) {
+        console.log(err.stack);
+
+        state = state.set('status', STATUS_ERROR);
+        state = state.addToHistory(
+          'A system error occurred during the process.'
+        );
+
+        return state;
+      })
+    );
+  };
+};
+
+// @returns Map[String, Promise[Map[String, *]]]
+Orchestrator.prototype.runDocumentDownloader = function (url, tempLocation) {
+  return new Map({
+    name: 'retrieve-resources',
+    promise: DocumentDownloader.fetchAndInstall(url, tempLocation)
+      .then(function () {
+        return new Map({
+          status: 'ok',
+          history: 'The file has been retrieved.'
+        });
+      }).catch(function (error) {
```

Comment:
all the subsequent run* function  have a very similar catch handler; probably some room for DRY